xbmc/filesystem/PluginDirectory.cpp

```diff
@@ -476,14 +476,28 @@ bool CPluginDirectory::WaitOnScriptResult(const std::string &scriptPath, int scr
 {
   bool cancelled = false;
 
-  if (!m_fetchComplete.WaitMSec(20))
+  // CPluginDirectory::GetDirectory can be called from the main and other threads.
+  // If called form the main thread, we need to bring up the BusyDialog in order to
+  // keep the render loop alive
+  if (g_application.IsCurrentThread())
+  {
+    if (!m_fetchComplete.WaitMSec(20))
+    {
+      CScriptObserver scriptObs(scriptId, m_fetchComplete);
+      if (!CGUIDialogBusy::WaitOnEvent(m_fetchComplete, 200))
+      {
+        cancelled = true;
+      }
+      scriptObs.Abort();
+    }
+  }
+  else
   {
-    CScriptObserver scriptObs(scriptId, m_fetchComplete);
-    if (!CGUIDialogBusy::WaitOnEvent(m_fetchComplete, 200))
+    // kill the script if it does not return within 10 seconds
+    if (!m_fetchComplete.WaitMSec(10000))
```

Comment:
Yatse use JSON to browse VFS from addons and timeout was at 15 second at start and there's was thousand of complains, it's now 30 sec and I still receive regularly complains. But IMO more than 30 sec is not user friendly, addon should manage data differently and request less data / use chunks.

10 seconds sounds really too short for most of the large addons like Youtube, Internet is slow by nature.